gui/journal.lua

```diff
@@ -0,0 +1,797 @@
+        else
+            self:eraseSelection()
+            self:insert(clipboard)
+        end
+
+    end
+end
+
+function TextEditorView:setText(text, cursor_x, cursor_y)
+    local changed = self.text ~= text
+    self.text = text
+    self:recomputeLines()
+
+    if cursor_x and cursor_y then
+        self:setCursor(cursor_x, cursor_y)
+    end
+
+    if changed and self.on_change then
+        self.on_change(text)
+    end
+end
+
+function TextEditorView:insert(text)
+    self:eraseSelection()
+    local index = self:cursorToIndex(
+        self.cursor.x - 1,
+        self.cursor.y
+    )
+
+    local new_text =
+        self.text:sub(1, index) ..
+        text ..
+        self.text:sub(index + 1)
+    self:setText(new_text, self.cursor.x + #text, self.cursor.y)
+end
+
+function TextEditorView:cursorToIndex(x, y)
+    local cursor = x
+    local lines = {table.unpack(self.lines, 1, y - 1)}
+    for _, line in ipairs(lines) do
+      cursor = cursor + #line
+    end
+
+    return cursor
+end
+
+function TextEditorView:indexToCursor(index)
+    for y, line in ipairs(self.lines) do
+        if index < #line then
+            return {x=index + 1, y=y}
+        end
+        index = index - #line
+    end
+
+    return {
+        x=#self.lines[#self.lines],
+        y=#self.lines
+    }
+end
+
+function TextEditorView:onRenderBody(dc)
+    dc:pen({fg=self.text_pen, bg=COLOR_RESET, bold=true})
+
+    local max_width = dc.width
+    local new_line = self.debug and NEWLINE or ''
+
+    for ind, line in ipairs(self.lines) do
+        -- do not render new lines symbol
+        local line = line:gsub(NEWLINE, new_line)
+        dc:string(line)
+        dc:newline()
+    end
+
+    local show_focus = not self:hasSelection()
+        and self.parent_view.focus
+        and gui.blink_visible(530)
+    if (show_focus) then
+        dc:seek(self.cursor.x - 1, self.cursor.y - 1)
+            :char('_')
+    end
+
+    if self:hasSelection() then
+        local sel_new_line = self.debug and PERIOD or ''
+        local from, to = self.cursor, self.sel_end
+        if (from.y > to.y or (from.y == to.y and from.x > to.x)) then
+            from, to = to, from
+        end
+
+        local line = self.lines[from.y]
+            :sub(from.x, to.y == from.y and to.x or nil)
+            :gsub(NEWLINE, sel_new_line)
+        dc:pen({ fg=self.text_pen, bg=self.select_pen })
+            :seek(from.x - 1, from.y - 1)
+            :string(line)
+
+        for y = from.y + 1, to.y - 1 do
+            line = self.lines[y]:gsub(NEWLINE, sel_new_line)
+            dc:seek(0, y - 1)
+                :string(line)
+        end
+
+        if (to.y > from.y) then
+            local line = self.lines[to.y]
+                :sub(1, to.x)
+                :gsub(NEWLINE, sel_new_line)
+            dc:seek(0, to.y - 1)
+                :string(line)
+        end
+
+        dc:pen({fg=self.text_pen, bg=COLOR_RESET})
+    end
+
+    if self.debug then
```

Comment:
consider a separate `widgets.Window` with debug state, including alternately formatted text. Example: https://github.com/DFHack/scripts/blob/master/gui/design.lua#L1509

Reply:
I would prefer to stay with current version for now, as its not crucial for most of work on the journal.
For sure I will consider it in next versions.